pkg/object/mqttproxy/topic.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+type topicMapFunc func(mqttTopic string) (topic string, headers map[string]string, err error)
+
+// TopicManager to manage topic subscribe and unsubscribe in MQTT
+type TopicManager struct {
+	sync.RWMutex
+	root *topicNode
+}
+
+func newTopicManager() *TopicManager {
+	return &TopicManager{
+		root: newNode(),
+	}
+}
+
+func validTopic(levels []string) bool {
+	for i, level := range levels {
+		if strings.ContainsRune(level, '+') && len(level) != 1 {
+			return false
+		}
+		if strings.ContainsRune(level, '#') {
+			if len(level) != 1 || i != len(levels)-1 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (mgr *TopicManager) subscribe(topics []string, qoss []byte, clientID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	for i, t := range topics {
+		levels := strings.Split(t, "/")
+		if !validTopic(levels) {
+			return fmt.Errorf("topic not valid, topic:%s", t)
+		}
+		if err := mgr.root.insert(levels, qoss[i], clientID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mgr *TopicManager) unsubscribe(topics []string, clientID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	for _, t := range topics {
+		levels := strings.Split(t, "/")
+		if !validTopic(levels) {
+			return fmt.Errorf("topic not valid, topic:%s", t)
+		}
```

Comment:
I've seen this snippet in many places. I think we can do two improvements here.
1. using a low-level function to remove the duplicated code - for example: `validate()`
2. using a low-level structure to maintain the `topic -> levels -> valid` relationship and state. 

so, every up-level functions (such as: validate, insert, remove, and match) just only operate the high-level data- `topic` instead of the low-level data - `levels[]`.

Reply:
Thanks a lot for your review, code efficiency is really important!